### PR TITLE
v1 refactors and model features

### PR DESCRIPTION
**Pull request by devinivy**
This includes a fair amount of cleanup/refactoring.  It also includes some missing model features covered by #4 and #5– using the joi schema to determine the attributes that should be serialized as JSON, and adjusting the joi schema so that it's suitable for `patch()`es (should not apply default values or enforce required fields).  Here's a summary,

1. Refactor out arbitrary random-id-ing of "knex-groups."  This was a big one!  I think we can further refactor the internal plugin state to be more ergonomic, and possibly more flexible (e.g. if in the future models within a single plugin might have different knex instances).
2. Handle memoization of joi schema (so that we don't compile joi schemas all the dang time... it's expensive).
3. Simplify model tests by not involving hapi when it's not necessary.
4. Support correct validation of `patch()`es (#4), with tests.
5. Support computing of (memoized) `jsonAttributes` (#5) based upon joi schema, with tests.
6. Do not rely on the internal version of joi for model validation.
7. No longer enforce lowercase model names.  PascalCase is actually desirable because our models are classes.
8. Miscellaneous cleanup and removing of dead/commented code.

### CODE REVIEW

File: test/index.js [hunk 1]
```diff
@@ -1221,58 +1115,170 @@ describe('Schwifty', () => {
 
                 done();
             });
-        });
 
-        it('skips validation if no schema exists on the model', (done) => {
+            it('skips validation when missing model is missing joiSchema.', (done) => {
```
Comment (wswoodruff) [inline test/index.js]: when model is missing joiSchema?

File: test/index.js [hunk 2]
```diff
@@ -1221,58 +1115,170 @@ describe('Schwifty', () => {
...
 
-                const NoSchemaClass = server.models().noschema;
+                expect(chompy.$validate(whateverSchema, { skipValidation: true })).to.equal(whateverSchema);
 
-                const anythingGoes = new NoSchemaClass();
+                done();
+            });
+
+            it('skips validation when `skipValidation` option is passed to $validate().', (done) => {
+
```
Comment (wswoodruff) [inline test/index.js:1134]: Duplicate test as the one right above it?

**Review by wswoodruff — COMMENTED** [review]
Looks great!!, only a couple minor comments. This thing is gonna rawx

File: test/index.js (hunk 1 above)
Comment (devinivy) [reply test/index.js]: That sounds right to me... maybe I'm missing something!

Comment (devinivy) [reply test/index.js]: OOPS

Comment (devinivy) [reply test/index.js]: LOL thank you

Comment (devinivy) [reply test/index.js]: IT SOUNDS FUNNY

File: test/index.js (hunk 2 above)
Comment (devinivy) [reply test/index.js:1134]: Good catch... wonder what I was going for there...

**Review by wswoodruff — APPROVED** [review]

**Comment by wswoodruff** [comment]
Whatsup with coveralls buggin out posting in triplicate?

**Pull request merged**